webrisk/database.go

```diff
@@ -0,0 +1,543 @@
+			db.setError(err)
+			// backoff strategy: MIN((2**N-1 * 15 minutes) * (RAND + 1), 24 hours)
+			n := 1 << db.updateAPIErrors
+			delay := time.Duration(float64(n) * (rand.Float64() + 1) * float64(baseRetryDelay))
+			if delay > maxRetryDelay {
+				delay = maxRetryDelay
+			}
+			db.updateAPIErrors++
+			return delay, false
+		}
+		resps = append(resps, resp)
+		if resp.RecommendedNextDiff != nil {
+			ndiff, _ :=pt.Timestamp(resp.RecommendedNextDiff)
+			serverMinWait := time.Duration(ndiff.Sub(time.Now()))
+			if serverMinWait > nextUpdateWait {
+				nextUpdateWait = serverMinWait
+				db.log.Printf("Server requested next update in %v", nextUpdateWait)
+			}
+		}
+	}
+
+
+	// If for some reason we missed a request or didn't get a response the
+	// rest of the logic may fail.
+	if len(s) != len(resps) {
+		db.setError(errors.New("mismatch between requests sent and responses received"))
+		return nextUpdateWait, false
+	}
+
+	db.updateAPIErrors = 0
+	// Update the threat database with the response.
+	db.generateThreatsForUpdate()
+	for i, resp := range resps {
+		// Assume a 1:1 correspondence between request and response
+		if err := db.tfu.update(resp, ThreatType(s[i].ThreatType)); err != nil {
+			db.setError(err)
+			db.log.Printf("update failure: %v", err)
+			db.tfu = nil
+			return nextUpdateWait, false
+		}
+	}
+
+	dbf := databaseFormat{make(threatsForUpdate), last}
+	for td, phs := range db.tfu {
+		// Copy of partialHashes before generateThreatsForLookups clobbers it.
+		dbf.Table[td] = phs
+	}
+
+	db.generateThreatsForLookups(last)
+
+	// Regenerate the database and store it.
+	if db.config.DBPath != "" {
+		// Semantically, we ignore save errors, but we do log them.
+		if err := saveDatabase(db.config.DBPath, dbf); err != nil {
+			db.log.Printf("save failure: %v", err)
+		}
+	}
+	return nextUpdateWait, true
+}
+
+// Lookup looks up the full hash in the threat list and returns a partial
+// hash and a set of ThreatTypes that may match the full hash.
+func (db *database) Lookup(hash hashPrefix) (h hashPrefix, tds []ThreatType) {
+	if !hash.IsFull() {
+		panic("hash is not full")
+	}
+
+	db.ml.RLock()
+	for td, hs := range db.tfl {
+		if n := hs.Lookup(hash); n > 0 {
+			h = hash[:n]
+			tds = append(tds, td)
+		}
+	}
+	db.ml.RUnlock()
+	return h, tds
+}
+
+// setError clears the database state and sets the last error to be err.
+//
+// This assumes that the db.mu lock is already held.
+func (db *database) setError(err error) {
+	db.tfu = nil
+
+	db.ml.Lock()
+	if db.err == nil {
+		db.readyCh = make(chan struct{})
+	}
+	db.tfl, db.err, db.last = nil, err, time.Time{}
+	db.ml.Unlock()
+}
+
+// isStale checks whether the last successful update should be considered stale.
+// Staleness is defined as being older than two of the configured update periods
+// plus jitter.
+func (db *database) isStale(lastUpdate time.Time) bool {
+	if db.config.now().Sub(lastUpdate) > 2*(db.config.UpdatePeriod+jitter) {
```

Comment:
Return this value directly, rather than wrapping in a `if condition { return true } return false`?

Reply:
Done.